Add `nexus` submodule with components for building customized NeXus file loaders

SimonHeybrock opened the pull request.

nvaytet commented on src/ess/nexus/_loader.py:
```diff
@@ -0,0 +1,159 @@
+    return schema(**loaded)
+
+
+T1 = TypeVar('T1', bound='InstrumentMixin')
```
Maybe better names for the types? Something that references Instrument and Entry instead of 1 and 2 ?

nvaytet commented on src/ess/nexus/_loader.py:
```diff
@@ -0,0 +1,159 @@
+    }
+
+
+def _select_events_and_load(detector, pulse_min=None, pulse_max=None, **kwargs):
```
Why is there a special function for pulses but not any of the other things you might want to filter by? (e.g. detector id)

Also, I don't see any test for `pulse_min` and `pulse_max`.

SimonHeybrock replied:
I could not think of a good case for supporting this by default. Filtering by detector ID is extremely inefficient (it has to load all the data anyway). Beyond that data may not be large? This can be added later or by the user if they have a case for it.

nvaytet commented on src/ess/nexus/_loader.py:
```diff
@@ -0,0 +1,159 @@
+
+Fields = make_section("Fields", [snx.Field, snx.NXlog])
+Choppers = make_section("Choppers", [snx.NXdisk_chopper, snx.NXfermi_chopper])
+Detectors = make_section("Detectors", snx.NXdetector, _select_events_and_load)
```
If you select pulses for the detector data, shouldn't you do the same for event-mode monitor data?

SimonHeybrock replied:
`scippnexus.NXmonitor` currently does not support the `events` property. Of course this could be added in the future.

nvaytet commented on src/ess/nexus/_loader.py:
```diff
@@ -0,0 +1,159 @@
+Source = make_leaf("Source", snx.NXsource)
+
+
+@dataclass
+class BasicInstrument(InstrumentMixin):
+    """
+    Basic loader component for NXinstrument.
+
+    Produces a basic subset of NXinstrument, loading NXdetector children and fields
+    (including NXlog). Subclass this in a new dataclass to add sections for more NeXus
+    classes.
+    """
+    fields: Fields
+    detectors: Detectors
+
+
+@dataclass
+class BasicEntry(EntryMixin):
+    """
+    Basic loader component for NXentry.
+
+    Produces a basic subset of NXentry, loading NXinstrument, NXmonitor, and NXsample
+    children as well as fields (including NXlog). Subclass this in a new dataclass to
+    add sections for more NeXus classes.
+    """
+    fields: Fields
+    instrument: BasicInstrument
+    monitors: Monitors
+    sample: Sample
```
:+1:

nvaytet commented on tests/nexus/load_test.py:
```diff
@@ -0,0 +1,132 @@
+
+
+@dataclass
+class InstrumentWithDetectorsPyPulse(nexus.InstrumentMixin):
```
```suggestion
class InstrumentWithDetectorsByPulse(nexus.InstrumentMixin):
```

nvaytet commented on tests/nexus/load_test.py:
```diff
@@ -0,0 +1,132 @@
+
+
+def test_load_instrument_without_binning_to_pixels(nxroot):
+    data = InstrumentWithDetectorsPyPulse.from_nexus(nxroot.entry)
```
```suggestion
    data = InstrumentWithDetectorsByPulse.from_nexus(nxroot.entry)
```

nvaytet commented on src/ess/nexus/_loader.py:
```diff
@@ -0,0 +1,159 @@
+    fields: Fields
+    instrument: BasicInstrument
+    monitors: Monitors
+    sample: Sample
```
Not loading the source?

SimonHeybrock replied:
This will need iterations. I don't know if `Source` always exists, or is needed?

nvaytet commented on src/ess/nexus/_loader.py:
```diff
@@ -0,0 +1,159 @@
+    """
+
+    def from_nexus(cls, group: snx.NXobject, **kwargs):
+        c = group[key] if isinstance(key, str) else group.__getattr__(key.__name__[2:])
```
Is the `key.__name__[2:]` to remove the `NX` prefix? Could this be brittle logic?

SimonHeybrock replied:
Yes and no. The `NX` prefix is standardized in Nexus. But of course one could pass a Python class that does not follow this schema. One could require providing the name as a separate argument, but I feel that would be *more* likely to break in practice?

nvaytet commented on tests/nexus/load_test.py:
```diff
@@ -0,0 +1,132 @@
+
+
+def test_loaded_named_leaf(nxroot):
+    data = NamedLeaf.from_nexus(nxroot.entry.instrument)
```
What happens if you try to use a `make_leaf` type on something that is not a leaf? (i.e. there is still some data nested below)

SimonHeybrock replied:
You will get an exception from scippnexus, complaining that this NX_class cannot be loaded.

jl-wynen commented on src/ess/nexus/_loader.py:
```diff
@@ -0,0 +1,159 @@
+        if isinstance(group, snx.NXroot):
+            return cls.from_nexus(group.entry, **kwargs)
+        with snx.File(group) as f:
+            return cls.from_nexus(f, **kwargs)
```
Why can this Mixin open a file but the others can't? I think it makes more sense to require the user to open it themselves in all cases.

SimonHeybrock replied:
@nvaytet @jl-wynen What are your thoughts on this? Would you prefer implementing a wrapper function that opens the file?

jl-wynen commented on src/ess/nexus/_loader.py:
```diff
@@ -0,0 +1,159 @@
+    except (snx.NexusStructureError, KeyError, sc.DTypeError, ValueError,
+            IndexError) as e:
+        if not skip_errors:
+            raise e from None
```
Just `raise` (without `e`) to get the full backtrace?

jl-wynen commented on src/ess/nexus/_loader.py:
```diff
@@ -0,0 +1,159 @@
+    def from_nexus(cls, group: snx.NXobject, **kwargs):
+        return cls(_load_multi(load, group[key], **kwargs))
+
+    return type(name, (dict, ), dict(from_nexus=classmethod(from_nexus)))
```
Not sure what implications this has. Is subclassing `dict` supported? Would it make more sense to use a dataclass or NamedDict? Or would those require knowing the possible fields ahead of time?

SimonHeybrock replied:
> Is subclassing dict supported?
Supported by what?

> Would it make more sense to use a dataclass or NamedDict? Or would those require knowing the possible fields ahead of time?
Indeed, we do not know all the field names (e.g., the names of all NXdetector in the file).

jl-wynen replied:
> >Is subclassing dict supported?
> Supported by what?

By dict and the rest of the ecosystem.

SimonHeybrock replied:
Yes, `dict` can be subclassed.

jl-wynen reviewed (commented):
For my own understanding: Is the reason for implementing this interface that it makes assumptions about the structure of a file, e.g. an instrument knows that it contains fields and detectors as direct children?
Is this why this code is in `ess` and not `scippnexus`?

SimonHeybrock commented:
> For my own understanding: Is the reason for implementing this interface that it makes assumptions about the structure of a file, e.g. an instrument knows that it contains fields and detectors as direct children?
> Is this why this code is in `ess` and not `scippnexus`?

I think we may eventually decide to move this into `scippnexus`. I did it here initially, since I think we will likely need some iterations, and want to try how it integrates with implementation of data reduction workflows. Having it in `scippnexus` would tie us more to releases and complicate this.

jl-wynen commented:
But then I am unclear about what benefit this has over `group[NXInstrument]`.

SimonHeybrock commented:
> But then I am unclear about what benefit this has over `group[NXInstrument]`.

You mean the additions in this PR? It avoids a significant amount of boilerplate code, I think.

jl-wynen commented:
> > But then I am unclear about what benefit this has over `group[NXInstrument]`.
> 
> You mean the additions in this PR? It avoids a significant amount of boilerplate code, I think.

But why not use `group[NXInstrument][()]` instead of `Instrument.from_nexus(group)`?

SimonHeybrock commented:
> > > But then I am unclear about what benefit this has over `group[NXInstrument]`.
> > 
> > 
> > You mean the additions in this PR? It avoids a significant amount of boilerplate code, I think.
> 
> But why not use `group[NXInstrument][()]` instead of `Instrument.from_nexus(group)`?

The former would necessarily load everything, whereas the approach implemented here allows for customization?

jl-wynen commented:
> The former would necessarily load everything, whereas the approach implemented here allows for customization?

Do you mean because the dataclasses here do not have to include all fields in a nexus group? If so, wouldn't this be specific to a use case?

SimonHeybrock commented:
> > The former would necessarily load everything, whereas the approach implemented here allows for customization?
> 
> Do you mean because the dataclasses here do not have to include all fields in a nexus group? If so, wouldn't this be specific to a use case?

Yes, the idea is that each instrument would customize this, e.g., to load (ore ignore) specific instrument components.

SimonHeybrock commented:
I finally addressed some of the comments here. See answers and updates. As mentioned face-to-face, I am certain that this will require iterations. For now, we can avoid documenting and advertising it widely such that we can try it in practice for a small subset of our applications.

SimonHeybrock commented:
I cannot reproduce this error locally. Tried both the latest `scippnexus` and 0.3.2. Can one of you reproduce this?

SimonHeybrock commented:
The CI fail comes from a NumPy Deprecation warning, triggered by and old h5py (h5py=3.1.0). This was fixed one and a half years ago in h5py=3.2.0. Can we update @nvaytet?

nvaytet commented:
Yes, Mantid is forcing us to pin I think, although I have not tried in a while, but `hdf5=1.10` is still required by the mantid conda recipe. See the discussion in Slack:
![Screenshot at 2022-11-14 12-23-54](https://user-images.githubusercontent.com/39047984/201648335-ce3e51cc-f5e7-460b-9c11-8c227e2f8f80.png)

SimonHeybrock commented:
So.... should we suppress the warning instead? Or can we avoid the Mantid dependency in the CI?

nvaytet commented:
Not sure, I've just asked on the Mantid slack if we could move `hdf5` to a more recent version. We'll see what happens.

SimonHeybrock commented:
I don't think a change there would happen on a timescale that works for us. I think we should begin considering dropping Mantid from the CI, or finding ways to split things.

nvaytet reviewed (approved):
At this point, I feel further discussion would not be the most productive. It looks ok to me, and I think we should just try it out in practise, this will tell us much quicker whether the approach needs substantial changes or not.
